From f8688d733bc1ecf86e5b7af7c40b758c0c9c4f28 Mon Sep 17 00:00:00 2001
From: Michael R Sweet <michael.r.sweet@gmail.com>
Date: Thu, 5 Sep 2019 12:43:58 -0400
Subject: [PATCH] Add workaround for systemd's lack of true launch-on-demand
 support (Issue #5640)

diff --git a/scheduler/main.c b/scheduler/main.c
index 842b00a42..d5fdf971a 100644
--- a/scheduler/main.c
+++ b/scheduler/main.c
@@ -757,7 +757,9 @@ main(int  argc,				/* I - Number of command-line args */
 #ifdef HAVE_ONDEMAND
 	if (OnDemand)
 	{
+#  ifndef HAVE_SYSTEMD /* Issue #5640: systemd doesn't actually support launch-on-demand services, need to fake it */
 	  stop_scheduler = 1;
+#  endif /* HAVE_SYSTEMD */
 	  break;
 	}
 #endif /* HAVE_ONDEMAND */
-- 
2.20.1

