From 5910c99b9211641e0691b2d418ee546af54c7a8c Mon Sep 17 00:00:00 2001
From: Michael R Sweet <msweet@msweet.org>
Date: Fri, 11 Apr 2025 10:40:38 -0400
Subject: [PATCH] Fix memory leak in httpClose (Issue #1223)

---
 CHANGES.md  | 9 ++++++++-
 cups/http.c | 8 +++++++-
 2 files changed, 15 insertions(+), 2 deletions(-)

diff --git a/cups/http.c b/cups/http.c
index 31a8be361..889760fd2 100644
--- a/cups/http.c
+++ b/cups/http.c
@@ -350,6 +350,7 @@ httpClearFields(http_t *http)		/* I - HTTP connection */
 void
 httpClose(http_t *http)			/* I - HTTP connection */
 {
+  http_field_t	field;			/* Current field */
 #ifdef HAVE_GSSAPI
   OM_uint32	minor_status;		/* Minor status code */
 #endif /* HAVE_GSSAPI */
@@ -392,7 +393,12 @@ httpClose(http_t *http)			/* I - HTTP connection */
     AuthorizationFree(http->auth_ref, kAuthorizationFlagDefaults);
 #endif /* HAVE_AUTHORIZATION_H */
 
-  httpClearFields(http);
+  for (field = HTTP_FIELD_ACCEPT_LANGUAGE; field < HTTP_FIELD_MAX; field ++)
+  {
+    free(http->default_fields[field]);
+    if (field >= HTTP_FIELD_ACCEPT_ENCODING || http->fields[field] != http->_fields[field])
+      free(http->fields[field]);
+  }
 
   if (http->authstring && http->authstring != http->_authstring)
     free(http->authstring);
-- 
2.49.0

