diff -up cups-1.3.5/cups/auth.c.peercred cups-1.3.5/cups/auth.c
--- cups-1.3.5/cups/auth.c.peercred	2008-02-05 16:50:17.000000000 +0000
+++ cups-1.3.5/cups/auth.c	2008-02-05 16:50:22.000000000 +0000
@@ -36,6 +36,7 @@
 #include <ctype.h>
 #include <errno.h>
 #include <fcntl.h>
+#include <sys/socket.h>
 #include <sys/stat.h>
 #if defined(WIN32) || defined(__EMX__)
 #  include <io.h>
diff -up cups-1.3.5/scheduler/auth.c.peercred cups-1.3.5/scheduler/auth.c
--- cups-1.3.5/scheduler/auth.c.peercred	2008-02-05 15:44:57.000000000 +0000
+++ cups-1.3.5/scheduler/auth.c	2008-02-05 16:50:36.000000000 +0000
@@ -86,6 +86,7 @@ extern const char *cssmErrorString(int e
 #ifdef HAVE_SYS_PARAM_H
 #  include <sys/param.h>
 #endif /* HAVE_SYS_PARAM_H */
+#include <sys/socket.h>
 #ifdef HAVE_SYS_UCRED_H
 #  include <sys/ucred.h>
 typedef struct xucred cupsd_ucred_t;
