diff -up cups-1.4b1/conf/mime.convs.in.paps cups-1.4b1/conf/mime.convs.in
--- cups-1.4b1/conf/mime.convs.in.paps	2008-10-08 22:07:45.000000000 +0100
+++ cups-1.4b1/conf/mime.convs.in	2008-11-11 15:42:18.000000000 +0000
@@ -45,7 +45,7 @@ application/x-cshell		application/postsc
 application/x-csource		application/postscript		33	texttops
 application/x-perl		application/postscript		33	texttops
 application/x-shell		application/postscript		33	texttops
-text/plain			application/postscript		33	texttops
+text/plain			application/postscript		33	texttopaps
 text/html			application/postscript		33	texttops
 image/gif			application/vnd.cups-postscript	66	imagetops
 image/png			application/vnd.cups-postscript	66	imagetops
