diff -up cups-1.3.7/backend/ipp.c.str2750 cups-1.3.7/backend/ipp.c
--- cups-1.3.7/backend/ipp.c.str2750	2008-06-03 16:07:31.000000000 +0100
+++ cups-1.3.7/backend/ipp.c	2008-06-03 16:07:43.000000000 +0100
@@ -1483,7 +1483,7 @@ password_cb(const char *prompt)		/* I - 
 {
   (void)prompt;
 
-  if (password && password_tries < 3)
+  if (password && *password && password_tries < 3)
   {
     password_tries ++;
 
