--- cups-1.2.1/Makedefs.in.relro	2006-07-04 17:52:05.000000000 +0100
+++ cups-1.2.1/Makedefs.in	2006-07-04 17:52:54.000000000 +0100
@@ -132,7 +132,7 @@
 IMGLIBS		=	@IMGLIBS@
 IMGFILTERS	=	@IMGFILTERS@
 LDFLAGS		=	-L../cups -L../filter $(ARCHFLAGS) \
-			@LDFLAGS@ @PIEFLAGS@ $(OPTIM)
+			@LDFLAGS@ @RELROFLAG@ @PIEFLAGS@ $(OPTIM)
 LINKCUPS	=	@LINKCUPS@ $(SSLLIBS)
 LINKCUPSIMAGE	=	@LINKCUPSIMAGE@
 LIBS		=	$(LINKCUPS) $(COMMONLIBS)
--- cups-1.2.1/config-scripts/cups-compiler.m4.relro	2006-07-04 17:52:05.000000000 +0100
+++ cups-1.2.1/config-scripts/cups-compiler.m4	2006-07-04 17:52:05.000000000 +0100
@@ -89,6 +89,9 @@
 CXXLIBS=""
 AC_SUBST(CXXLIBS)
 
+RELROFLAG=""
+AC_SUBST(RELROFLAG)
+
 PIEFLAGS=""
 AC_SUBST(PIEFLAGS)
 
@@ -108,6 +111,12 @@
 
 	case $uname in
 		Linux*)
+			RELROFLAG="-Wl,-z,relro"
+			;;
+	esac
+
+	case $uname in
+		Linux*)
 			if test x$enable_pie = xyes; then
 				PIEFLAGS="-pie -fPIE"
 			fi
