diff -up cups-2.2.12/scheduler/cups.service.in.ypbind cups-2.2.12/scheduler/cups.service.in
--- cups-2.2.12/scheduler/cups.service.in.ypbind	2019-08-19 10:00:47.586326493 +0200
+++ cups-2.2.12/scheduler/cups.service.in	2019-08-19 10:01:39.295890076 +0200
@@ -1,7 +1,7 @@
 [Unit]
 Description=CUPS Scheduler
 Documentation=man:cupsd(8)
-After=network.target sssd.service
+After=network.target sssd.service ypbind.service
 
 [Service]
 ExecStart=@sbindir@/cupsd -l
