diff -up cups-1.4b2/CHANGES-1.3.txt.str3059 cups-1.4b2/CHANGES-1.3.txt
diff -up cups-1.4b2/scheduler/type.c.str3059 cups-1.4b2/scheduler/type.c
--- cups-1.4b2/scheduler/type.c.str3059	2008-07-11 23:46:21.000000000 +0100
+++ cups-1.4b2/scheduler/type.c	2009-01-28 17:40:23.000000000 +0000
@@ -188,8 +188,7 @@ mimeAddTypeRule(mime_type_t *mt,	/* I - 
     else if (*rule == '+' && current != NULL)
     {
       if (logic != MIME_MAGIC_AND &&
-          current != NULL && current->prev != NULL &&
-	  current->prev->prev != NULL)
+          current != NULL && current->prev != NULL)
       {
        /*
         * OK, we have more than 1 rule in the current tree level...  Make a
