diff --git a/conf/pam.password b/conf/pam.password
new file mode 100644
index 000000000..6146a9134
--- /dev/null
+++ b/conf/pam.password
@@ -0,0 +1,4 @@
+#%PAM-1.0
+# Use password-auth common PAM configuration for the daemon
+auth     include  password-auth
+account  include  password-auth
diff --git a/conf/pam.system b/conf/pam.system
new file mode 100644
index 000000000..9e7c820f4
--- /dev/null
+++ b/conf/pam.system
@@ -0,0 +1,3 @@
+#%PAM-1.0
+auth     include  system-auth
+account  include  system-auth
diff --git a/config-scripts/cups-pam.m4 b/config-scripts/cups-pam.m4
index cf983b031..f4d6ecbd9 100644
--- a/config-scripts/cups-pam.m4
+++ b/config-scripts/cups-pam.m4
@@ -71,6 +71,10 @@ AS_IF([test x$enable_pam != xno], [
 	    PAMMOD="pam_${with_pam_module}.so"
 	], [test -f /etc/pam.d/common-auth], [
 	    PAMFILE="pam.common"
+	], [test -f /etc/pam.d/password-auth], [
+	    PAMFILE="pam.password"
+	], [test -f /etc/pam.d/system-auth], [
+	    PAMFILE="pam.system"
 	], [
 	    moddir=""
 	    for dir in /lib/security /lib64/security /lib/x86_64-linux-gnu/security /var/lib/pam; do
diff --git a/configure b/configure
index 7af3d1c9c..1921738cf 100755
--- a/configure
+++ b/configure
@@ -10396,6 +10396,16 @@ then :
 
 	    PAMFILE="pam.common"
 
+elif test -f /etc/pam.d/password-auth
+then :
+
+	    PAMFILE="pam.password"
+
+elif test -f /etc/pam.d/system-auth
+then :
+
+	    PAMFILE="pam.system"
+
 else $as_nop
 
 	    moddir=""
