diff --git a/cups/ppd-cache.c b/cups/ppd-cache.c
index 40881cd39..86f54060b 100644
--- a/cups/ppd-cache.c
+++ b/cups/ppd-cache.c
@@ -3441,6 +3441,27 @@ _ppdCreateFromIPP2(
     cupsFilePrintf(fp, "*cupsJobPassword: \"%s\"\n", pattern);
   }
 
+
+ /*
+  * Document Password printing...
+  * CUPS doesn't actually have a cupsDocumentPassword PPD attribute, we just
+  * use this custom attribute to indicate the printer supports document-password
+  */
+
+  if ((attr = ippFindAttribute(supported, "document-password-supported", IPP_TAG_INTEGER)) != NULL)
+  {
+    int		maxlen = ippGetInteger(attr, 0);
+					/* Maximum length */
+
+    if (maxlen > 1023)
+      maxlen = 1023;
+
+    if (maxlen < 255)
+      maxlen = 255;
+
+    cupsFilePrintf(fp, "*cupsDocumentPassword: \"%i\"\n", maxlen);
+  }
+
  /*
   * Filters...
   */
diff --git a/scheduler/printers.c b/scheduler/printers.c
index 5f9852e64..2f11edfa1 100644
--- a/scheduler/printers.c
+++ b/scheduler/printers.c
@@ -4083,6 +4083,19 @@ load_ppd(cupsd_printer_t *p)		/* I - Printer */
                     "job-password-supported", (int)strlen(p->pc->password));
     }
 
+    if ((ppd_attr = ppdFindAttr(ppd, "cupsDocumentPassword", NULL)) != NULL)
+
+    {
+      ippAddString(p->ppd_attrs, IPP_TAG_PRINTER, IPP_TAG_KEYWORD,
+                   "job-password-encryption-supported", NULL, "none");
+      ippAddInteger(p->ppd_attrs, IPP_TAG_PRINTER, IPP_TAG_INTEGER,
+                    "job-password-supported", 4);
+      ippAddString(p->ppd_attrs, IPP_TAG_PRINTER, IPP_TAG_KEYWORD,
+                   "job-password-repertoire-configured", NULL, "iana_utf-8_digits");
+      ippAddInteger(p->ppd_attrs, IPP_TAG_PRINTER, IPP_TAG_INTEGER,
+                    "document-password-supported", ippGetInteger(ppd_attr, 0));
+    }
+
     if (ppd->throughput)
     {
       ippAddInteger(p->ppd_attrs, IPP_TAG_PRINTER, IPP_TAG_INTEGER,
