diff -up cups-1.3.6/cups/encode.c.str2715 cups-1.3.6/cups/encode.c
--- cups-1.3.6/cups/encode.c.str2715	2008-01-28 23:10:10.000000000 +0000
+++ cups-1.3.6/cups/encode.c	2008-02-23 12:16:50.000000000 +0000
@@ -67,6 +67,8 @@ static const _ipp_option_t ipp_options[]
   { 0, "job-page-limit",	IPP_TAG_INTEGER,	IPP_TAG_JOB },
   { 0, "job-priority",		IPP_TAG_INTEGER,	IPP_TAG_JOB },
   { 0, "job-quota-period",	IPP_TAG_INTEGER,	IPP_TAG_JOB },
+  { 1, "job-sheets",		IPP_TAG_NAME,		IPP_TAG_JOB },
+  { 1, "job-sheets-default",	IPP_TAG_NAME,		IPP_TAG_PRINTER },
   { 0, "job-uuid",		IPP_TAG_URI,		IPP_TAG_JOB },
   { 0, "landscape",		IPP_TAG_BOOLEAN,	IPP_TAG_JOB },
   { 1, "media",			IPP_TAG_KEYWORD,	IPP_TAG_JOB },
