diff -up cups-1.4rc1/cups/request.c.str3311 cups-1.4rc1/cups/request.c
--- cups-1.4rc1/cups/request.c.str3311	2009-08-27 13:16:17.965400634 +0100
+++ cups-1.4rc1/cups/request.c	2009-08-27 13:16:18.247275208 +0100
@@ -266,7 +266,9 @@ cupsDoIORequest(http_t     *http,	/* I -
 
     DEBUG_printf(("2cupsDoIORequest: status=%d", status));
 
-    if (status == HTTP_FORBIDDEN || status >= HTTP_SERVER_ERROR)
+    if (status >= HTTP_BAD_REQUEST &&
+	status != HTTP_UNAUTHORIZED &&
+	status != HTTP_UPGRADE_REQUIRED)
     {
       httpFlush(http);
       _cupsSetHTTPError(status);
