diff -up cups-1.4b2-svn8404/scheduler/cups-driverd.cxx.str3124 cups-1.4b2-svn8404/scheduler/cups-driverd.cxx
--- cups-1.4b2-svn8404/scheduler/cups-driverd.cxx.str3124	2009-02-19 17:56:47.000000000 +0000
+++ cups-1.4b2-svn8404/scheduler/cups-driverd.cxx	2009-03-10 17:46:05.000000000 +0000
@@ -1716,7 +1716,7 @@ load_ppds(const char *d,		/* I - Actual 
     }
 
     for (i = 0; i < (int)(sizeof(languages) / sizeof(languages[0])); i ++)
-      if (!strcmp(languages[i].version, lang_version))
+      if (!strcasecmp(languages[i].version, lang_version))
         break;
 
     if (i < (int)(sizeof(languages) / sizeof(languages[0])))
@@ -1731,10 +1731,10 @@ load_ppds(const char *d,		/* I - Actual 
     else
     {
      /*
-      * Unknown language; use "xx"...
+      * Unknown language; use "C"...
       */
 
-      strcpy(lang_version, "xx");
+      strcpy(lang_version, "C");
     }
 
    /*
