diff -up ./postscript.ppd.generic-ps ./postscript.ppd
--- ./postscript.ppd.generic-ps	2009-10-08 11:08:07.389336611 +0100
+++ ./postscript.ppd	2009-10-08 11:06:40.836338000 +0100
@@ -12,11 +12,11 @@
 *LanguageVersion: English
 *LanguageEncoding: ISOLatin1
 *PCFileName:	"POSTSCRIPT.PPD"
-*Manufacturer:	"Postscript"
-*Product:	"(CUPS v1.1)"
-*ModelName:     "Generic postscript printer"
-*ShortNickName: "Generic postscript printer"
-*NickName:      "Generic postscript printer"
+*Manufacturer:	"Generic"
+*Product:	"(PostScript Printer)"
+*ModelName:     "Generic PostScript Printer"
+*ShortNickName: "Generic PostScript Printer"
+*NickName:      "Generic PostScript Printer"
 *PSVersion:	"(2017.000) 0"
 *LanguageLevel:	"2"
 *ColorDevice:	False
