diff --git a/backend/ipp.c b/backend/ipp.c
index be1f2c048..e87864695 100644
--- a/backend/ipp.c
+++ b/backend/ipp.c
@@ -272,7 +272,7 @@ main(int  argc,				/* I - Number of command-line args */
   ppd_file_t	*ppd = NULL;		/* PPD file */
   _ppd_cache_t	*pc = NULL;		/* PPD cache and mapping data */
   fd_set	input;			/* Input set for select() */
-
+  const char	*doc_password;		/* document-password */
 
  /*
   * Make sure status messages are not buffered...
@@ -1327,10 +1327,20 @@ main(int  argc,				/* I - Number of command-line args */
     }
 
    /*
-    * Validate job-password/-encryption...
+    * Validate document-password, job-password/-encryption...
     */
-
-    if (cupsGetOption("job-password", num_options, options))
+    doc_password = NULL;
+    if ((cupsGetOption("document-password", num_options, options)) != NULL) {
+        cupsSetEncryption(HTTP_ENCRYPTION_ALWAYS);
+    } else if ((doc_password = cupsGetOption("job-password", num_options, options)) != NULL && ppdFindAttr(ppd, "cupsJobPassword", NULL) == NULL)
+    {
+      /*
+       * we map job-password to document-password for printers that don't support job-password
+       */
+      num_options = cupsAddOption("document-password", doc_password, num_options, &options);
+      num_options = cupsRemoveOption("job-password", num_options, &options);
+      cupsSetEncryption(HTTP_ENCRYPTION_ALWAYS);
+    } else if (cupsGetOption("job-password", num_options, options))
     {
       const char *keyword;		/* job-password-encryption value */
       static const char * const hashes[] =
diff --git a/cups/ipp-support.c b/cups/ipp-support.c
index bb0c6c8bd..51d5edddd 100644
--- a/cups/ipp-support.c
+++ b/cups/ipp-support.c
@@ -1670,6 +1670,7 @@ ippCreateRequestedArray(ipp_t *request)	/* I - IPP request */
     "job-media-sheets-supported",
     "job-page-limit",			/* CUPS extension */
     "job-password-encryption-supported",
+    "job-password-repertoire-configured",
     "job-password-supported",
     "job-presets-supported",		/* IPP Presets */
     "job-privacy-attributes",		/* IPP Privacy Attributes */
diff --git a/cups/ppd-cache.c b/cups/ppd-cache.c
index 40881cd39..86f54060b 100644
--- a/cups/ppd-cache.c
+++ b/cups/ppd-cache.c
@@ -3441,6 +3441,27 @@ _ppdCreateFromIPP2(
     cupsFilePrintf(fp, "*cupsJobPassword: \"%s\"\n", pattern);
   }
 
+
+ /*
+  * Document Password printing...
+  * CUPS doesn't actually have a cupsDocumentPassword PPD attribute, we just
+  * use this custom attribute to indicate the printer supports document-password
+  */
+
+  if ((attr = ippFindAttribute(supported, "document-password-supported", IPP_TAG_INTEGER)) != NULL)
+  {
+    int		maxlen = ippGetInteger(attr, 0);
+					/* Maximum length */
+
+    if (maxlen > 1023)
+      maxlen = 1023;
+
+    if (maxlen < 255)
+      maxlen = 255;
+
+    cupsFilePrintf(fp, "*cupsDocumentPassword: \"%i\"\n", maxlen);
+  }
+
  /*
   * Filters...
   */
diff --git a/scheduler/printers.c b/scheduler/printers.c
index 5f9852e64..8c1dadf74 100644
--- a/scheduler/printers.c
+++ b/scheduler/printers.c
@@ -4083,6 +4083,24 @@ load_ppd(cupsd_printer_t *p)		/* I - Printer */
                     "job-password-supported", (int)strlen(p->pc->password));
     }
 
+    /* macOS and Windows print dialogs don't have support for document-password
+       but do for job-password, so we map document-password to job-password and
+       elsewhere vice-versa for printers that support document-password but
+       not job-password */
+    if ((ppd_attr = ppdFindAttr(ppd, "cupsDocumentPassword", NULL)) != NULL)
+    {
+      if (!p->pc || !(p->pc && p->pc->password)) {
+        ippAddString(p->ppd_attrs, IPP_TAG_PRINTER, IPP_TAG_KEYWORD,
+                     "job-password-encryption-supported", NULL, "none");
+        ippAddInteger(p->ppd_attrs, IPP_TAG_PRINTER, IPP_TAG_INTEGER,
+                      "job-password-supported", 4); /* 4 digit PIN */
+        ippAddString(p->ppd_attrs, IPP_TAG_PRINTER, IPP_TAG_KEYWORD,
+                     "job-password-repertoire-configured", NULL, "iana_utf-8_digits");
+      }
+      ippAddInteger(p->ppd_attrs, IPP_TAG_PRINTER, IPP_TAG_INTEGER,
+                    "document-password-supported", atoi(ppd_attr->value));
+    }
+
     if (ppd->throughput)
     {
       ippAddInteger(p->ppd_attrs, IPP_TAG_PRINTER, IPP_TAG_INTEGER,
