diff -up cups-1.3.7/scheduler/conf.c.serverbin-compat cups-1.3.7/scheduler/conf.c
--- cups-1.3.7/scheduler/conf.c.serverbin-compat	2008-03-20 04:06:01.000000000 +0000
+++ cups-1.3.7/scheduler/conf.c	2008-07-15 11:04:19.000000000 +0100
@@ -402,6 +402,9 @@ cupsdReadConfiguration(void)
   cupsdSetString(&ServerName, httpGetHostname(NULL, temp, sizeof(temp)));
   cupsdSetStringf(&ServerAdmin, "root@%s", temp);
   cupsdSetString(&ServerBin, CUPS_SERVERBIN);
+#ifdef __x86_64__
+  cupsdSetString(&ServerBin_compat, "/usr/lib64/cups");
+#endif /* __x86_64__ */
   cupsdSetString(&RequestRoot, CUPS_REQUESTS);
   cupsdSetString(&CacheDir, CUPS_CACHEDIR);
   cupsdSetString(&DataDir, CUPS_DATADIR);
@@ -1095,7 +1098,12 @@ cupsdReadConfiguration(void)
     * Read the MIME type and conversion database...
     */
 
+#ifdef __x86_64__
+    snprintf(temp, sizeof(temp), "%s/filter:%s/filter", ServerBin,
+	     ServerBin_compat);
+#else /* ! defined(__x86_64__) */
     snprintf(temp, sizeof(temp), "%s/filter", ServerBin);
+#endif
 
     MimeDatabase = mimeLoad(ServerRoot, temp);
 
diff -up cups-1.3.7/scheduler/conf.h.serverbin-compat cups-1.3.7/scheduler/conf.h
--- cups-1.3.7/scheduler/conf.h.serverbin-compat	2008-02-15 23:26:51.000000000 +0000
+++ cups-1.3.7/scheduler/conf.h	2008-07-15 11:04:19.000000000 +0100
@@ -59,6 +59,10 @@ VAR char		*ConfigurationFile	VALUE(NULL)
 					/* Root directory for scheduler */
 			*ServerBin		VALUE(NULL),
 					/* Root directory for binaries */
+#ifdef __x86_64__
+			*ServerBin_compat	VALUE(NULL),
+					/* Compat directory for binaries */
+#endif /* __x86_64__ */
 			*StateDir		VALUE(NULL),
 					/* Root directory for state data */
 			*RequestRoot		VALUE(NULL),
diff -up cups-1.3.7/scheduler/env.c.serverbin-compat cups-1.3.7/scheduler/env.c
--- cups-1.3.7/scheduler/env.c.serverbin-compat	2007-07-11 22:46:42.000000000 +0100
+++ cups-1.3.7/scheduler/env.c	2008-07-15 11:04:19.000000000 +0100
@@ -86,8 +86,13 @@ cupsdInitEnv(void)
   cupsdSetEnv("LD_LIBRARY_PATH", NULL);
   cupsdSetEnv("LD_PRELOAD", NULL);
   cupsdSetEnv("NLSPATH", NULL);
+#ifdef __x86_64__
+  cupsdSetEnvf("PATH", "%s/filter:%s/filter:" CUPS_BINDIR ":" CUPS_SBINDIR
+                       ":/bin:/usr/bin", ServerBin, ServerBin_compat);
+#else /* ! defined(__x86_64__) */
   cupsdSetEnvf("PATH", "%s/filter:" CUPS_BINDIR ":" CUPS_SBINDIR
                        ":/bin:/usr/bin", ServerBin);
+#endif
   cupsdSetEnv("SERVER_ADMIN", ServerAdmin);
   cupsdSetEnv("SHLIB_PATH", NULL);
   cupsdSetEnv("SOFTWARE", CUPS_MINIMAL);
diff -up cups-1.3.7/scheduler/ipp.c.serverbin-compat cups-1.3.7/scheduler/ipp.c
--- cups-1.3.7/scheduler/ipp.c.serverbin-compat	2008-03-24 22:34:44.000000000 +0000
+++ cups-1.3.7/scheduler/ipp.c	2008-07-15 11:04:19.000000000 +0100
@@ -2377,9 +2377,18 @@ add_printer(cupsd_client_t  *con,	/* I -
         * Could not find device in list!
 	*/
 
+#ifdef __x86_64__
+       snprintf(srcfile, sizeof(srcfile), "%s/backend/%s", ServerBin_compat,
+		 scheme);
+	if (access(srcfile, X_OK))
+	  {
+#endif /* __x86_64__ */
 	send_ipp_status(con, IPP_NOT_POSSIBLE, _("Bad device-uri \"%s\"!"),
         	        attr->values[0].string.text);
 	return;
+#ifdef __x86_64__
+	  }
+#endif /* __x86_64__ */
       }
     }
 
diff -up cups-1.3.7/scheduler/job.c.serverbin-compat cups-1.3.7/scheduler/job.c
--- cups-1.3.7/scheduler/job.c.serverbin-compat	2008-03-25 22:57:40.000000000 +0000
+++ cups-1.3.7/scheduler/job.c	2008-07-15 11:04:19.000000000 +0100
@@ -3223,8 +3223,32 @@ start_job(cupsd_job_t     *job,		/* I - 
        i ++, filter = (mime_filter_t *)cupsArrayNext(filters))
   {
     if (filter->filter[0] != '/')
-      snprintf(command, sizeof(command), "%s/filter/%s", ServerBin,
-               filter->filter);
+      {
+	snprintf(command, sizeof(command), "%s/filter/%s", ServerBin,
+		 filter->filter);
+#ifdef __x86_64__
+	if (access(command, F_OK))
+	  {
+	    snprintf(command, sizeof(command), "%s/filter/%s",
+		     ServerBin_compat, filter->filter);
+	    if (!access(command, F_OK))
+	      {
+		/* Not in the correct directory, but found it in the compat
+		 * directory.  Issue a warning. */
+		cupsdLogMessage(CUPSD_LOG_INFO,
+				"Filter '%s' not in %s/filter!",
+				filter->filter, ServerBin);
+	      }
+	    else
+	      {
+		/* Not in the compat directory either; make any error
+		 * messages use the correct directory name then. */
+		snprintf(command, sizeof(command), "%s/filter/%s", ServerBin,
+			 filter->filter);
+	      }
+	  }
+#endif /* __x86_64__ */
+      }
     else
       strlcpy(command, filter->filter, sizeof(command));
 
@@ -3374,6 +3398,28 @@ start_job(cupsd_job_t     *job,		/* I - 
     {
       sscanf(printer->device_uri, "%254[^:]", method);
       snprintf(command, sizeof(command), "%s/backend/%s", ServerBin, method);
+#ifdef __x86_64__
+      if (access(command, F_OK))
+	{
+	  snprintf(command, sizeof(command), "%s/backend/%s", ServerBin_compat,
+		   method);
+	  if (!access(command, F_OK))
+	    {
+	      /* Not in the correct directory, but we found it in the compat
+	       * directory.  Issue a warning. */
+	      cupsdLogMessage(CUPSD_LOG_INFO,
+			      "Backend '%s' not in %s/backend!", method,
+			      ServerBin);
+	    }
+	  else
+	    {
+	      /* Not in the compat directory either; make any error
+		 messages use the correct directory name then. */
+	      snprintf(command, sizeof(command), "%s/backend/%s", ServerBin,
+		       method);
+	    }
+	}
+#endif /* __x86_64__ */
 
      /*
       * See if the backend needs to run as root...
diff -up cups-1.3.7/scheduler/printers.c.serverbin-compat cups-1.3.7/scheduler/printers.c
--- cups-1.3.7/scheduler/printers.c.serverbin-compat	2008-01-30 06:09:39.000000000 +0000
+++ cups-1.3.7/scheduler/printers.c	2008-07-15 11:04:19.000000000 +0100
@@ -934,9 +934,19 @@ cupsdLoadAllPrinters(void)
 	    * Backend does not exist, stop printer...
 	    */
 
+#ifdef __x86_64__
+	    snprintf(line, sizeof(line), "%s/backend/%s", ServerBin_compat,
+		     p->device_uri);
+	    if (access(line, F_OK))
+	      {
+#endif /* __x86_64__ */
+
 	    p->state = IPP_PRINTER_STOPPED;
 	    snprintf(p->state_message, sizeof(p->state_message),
 	             "Backend %s does not exist!", line);
+#ifdef __x86_64__
+	      }
+#endif /* __x86_64__ */
 	  }
         }
 
@@ -3395,6 +3405,12 @@ add_printer_filter(
 
     if (access(filename, X_OK))
     {
+#ifdef __x86_64__
+      snprintf(filename, sizeof(filename), "%s/filter/%s", ServerBin_compat,
+	       program);
+      if (access(filename, X_OK))
+	{
+#endif /* __x86_64__ */
       snprintf(p->state_message, sizeof(p->state_message),
                "Filter \"%s\" for printer \"%s\" not available: %s",
 	       program, p->name, strerror(errno));
@@ -3402,6 +3418,9 @@ add_printer_filter(
       cupsdSetPrinterState(p, IPP_PRINTER_STOPPED, 0);
 
       cupsdLogMessage(CUPSD_LOG_ERROR, "%s", p->state_message);
+#ifdef __x86_64__
+	}
+#endif /* __x86_64__ */
     }
   }
 
