diff -up cups-1.4rc1/cups/dest.c.str3285 cups-1.4rc1/cups/dest.c
--- cups-1.4rc1/cups/dest.c.str3285	2009-08-10 11:32:08.397714492 +0100
+++ cups-1.4rc1/cups/dest.c	2009-08-10 11:32:14.143839210 +0100
@@ -549,7 +549,7 @@ cupsGetNamedDest(http_t     *http,	/* I 
 
   if (!cups_get_sdests(http, op, name, 0, &dest))
   {
-    if (op == CUPS_GET_DEFAULT)
+    if (op == CUPS_GET_DEFAULT || name != NULL)
       return (NULL);
 
    /*
@@ -557,7 +557,7 @@ cupsGetNamedDest(http_t     *http,	/* I 
     * configuration file does not exist.  Find out the real default.
     */
 
-    if (!cups_get_sdests(http, CUPS_GET_DEFAULT, name, 0, &dest))
+    if (!cups_get_sdests(http, CUPS_GET_DEFAULT, NULL, 0, &dest))
       return (NULL);
   }
 
