diff -up cups-1.4rc1/backend/dnssd.c.str3272 cups-1.4rc1/backend/dnssd.c
--- cups-1.4rc1/backend/dnssd.c.str3272	2009-04-14 19:14:02.000000000 +0100
+++ cups-1.4rc1/backend/dnssd.c	2009-07-29 13:56:55.643396809 +0100
@@ -153,8 +153,8 @@ main(int  argc,				/* I - Number of comm
   memset(&action, 0, sizeof(action));
 
   sigemptyset(&action.sa_mask);
-  action.sa_handler = SIG_IGN;
-  sigaction(SIGTERM, &action, sigterm_handler);
+  action.sa_handler = sigterm_handler;
+  sigaction(SIGTERM, &action, NULL);
 #else
   signal(SIGTERM, sigterm_handler);
 #endif /* HAVE_SIGSET */
