diff -up cups-1.4rc1/templates/add-printer.tmpl.str3229 cups-1.4rc1/templates/add-printer.tmpl
--- cups-1.4rc1/templates/add-printer.tmpl.str3229	2008-09-20 04:33:59.000000000 +0100
+++ cups-1.4rc1/templates/add-printer.tmpl	2009-06-18 18:58:12.242510800 +0100
@@ -29,7 +29,7 @@
 </TR>
 <TR>
 <TH CLASS="label">Sharing:</TH>
-<TD><INPUT TYPE="CHECKBOX" NAME="PRINTER_IS_SHARED" {PRINTER_IS_SHARED=1?CHECKED:}">
+<TD><INPUT TYPE="CHECKBOX" NAME="PRINTER_IS_SHARED" {PRINTER_IS_SHARED=1?CHECKED:}>
 Share This Printer</TD>
 </TR>
 <TR>
diff -up cups-1.4rc1/templates/de/add-printer.tmpl.str3229 cups-1.4rc1/templates/de/add-printer.tmpl
--- cups-1.4rc1/templates/de/add-printer.tmpl.str3229	2009-04-03 16:55:28.000000000 +0100
+++ cups-1.4rc1/templates/de/add-printer.tmpl	2009-06-18 18:58:12.255510590 +0100
@@ -29,7 +29,7 @@
 </TR>
 <TR>
 <TH CLASS="label">Freigabe:</TH>
-<TD><INPUT TYPE="CHECKBOX" NAME="PRINTER_IS_SHARED" {PRINTER_IS_SHARED=1?CHECKED:}">
+<TD><INPUT TYPE="CHECKBOX" NAME="PRINTER_IS_SHARED" {PRINTER_IS_SHARED=1?CHECKED:}>
 Diesen Drucker freigeben</TD>
 </TR>
 <TR>
diff -up cups-1.4rc1/templates/de/modify-printer.tmpl.str3229 cups-1.4rc1/templates/de/modify-printer.tmpl
--- cups-1.4rc1/templates/de/modify-printer.tmpl.str3229	2009-04-03 16:55:28.000000000 +0100
+++ cups-1.4rc1/templates/de/modify-printer.tmpl	2009-06-18 18:58:12.258635708 +0100
@@ -23,7 +23,7 @@
 </TR>
 <TR>
 <TH CLASS="label">Freigabe:</TH>
-<TD><INPUT TYPE="CHECKBOX" NAME="PRINTER_IS_SHARED" {PRINTER_IS_SHARED=1?CHECKED:}">
+<TD><INPUT TYPE="CHECKBOX" NAME="PRINTER_IS_SHARED" {PRINTER_IS_SHARED=1?CHECKED:}>
 Diesen Drucker freigeben</TD>
 </TR>
 <TR>
diff -up cups-1.4rc1/templates/es/add-printer.tmpl.str3229 cups-1.4rc1/templates/es/add-printer.tmpl
--- cups-1.4rc1/templates/es/add-printer.tmpl.str3229	2008-12-08 21:20:06.000000000 +0000
+++ cups-1.4rc1/templates/es/add-printer.tmpl	2009-06-18 18:58:12.267511070 +0100
@@ -29,7 +29,7 @@
 </TR>
 <TR>
 <TH CLASS="label">Compartici&oacute;n:</TH>
-<TD><INPUT TYPE="CHECKBOX" NAME="PRINTER_IS_SHARED" {PRINTER_IS_SHARED=1?CHECKED:}">
+<TD><INPUT TYPE="CHECKBOX" NAME="PRINTER_IS_SHARED" {PRINTER_IS_SHARED=1?CHECKED:}>
 Compartir esta impresora</TD>
 </TR>
 <TR>
diff -up cups-1.4rc1/templates/es/modify-printer.tmpl.str3229 cups-1.4rc1/templates/es/modify-printer.tmpl
--- cups-1.4rc1/templates/es/modify-printer.tmpl.str3229	2008-12-08 21:20:06.000000000 +0000
+++ cups-1.4rc1/templates/es/modify-printer.tmpl	2009-06-18 18:58:12.270511552 +0100
@@ -23,7 +23,7 @@
 </TR>
 <TR>
 <TH CLASS="label">Compartida:</TH>
-<TD><INPUT TYPE="CHECKBOX" NAME="PRINTER_IS_SHARED" {PRINTER_IS_SHARED=1?CHECKED:}">
+<TD><INPUT TYPE="CHECKBOX" NAME="PRINTER_IS_SHARED" {PRINTER_IS_SHARED=1?CHECKED:}>
 Compartir esta impresora</TD>
 </TR>
 <TR>
diff -up cups-1.4rc1/templates/ja/add-printer.tmpl.str3229 cups-1.4rc1/templates/ja/add-printer.tmpl
--- cups-1.4rc1/templates/ja/add-printer.tmpl.str3229	2009-02-17 19:04:25.000000000 +0000
+++ cups-1.4rc1/templates/ja/add-printer.tmpl	2009-06-18 18:58:12.242510800 +0100
@@ -29,7 +29,7 @@
 </TR>
 <TR>
 <TH CLASS="label">共有:</TH>
-<TD><INPUT TYPE="CHECKBOX" NAME="PRINTER_IS_SHARED" {PRINTER_IS_SHARED=1?CHECKED:}">
+<TD><INPUT TYPE="CHECKBOX" NAME="PRINTER_IS_SHARED" {PRINTER_IS_SHARED=1?CHECKED:}>
 このプリンターを共有する</TD>
 </TR>
 <TR>
diff -up cups-1.4rc1/templates/ja/modify-printer.tmpl.str3229 cups-1.4rc1/templates/ja/modify-printer.tmpl
--- cups-1.4rc1/templates/ja/modify-printer.tmpl.str3229	2009-02-17 19:04:25.000000000 +0000
+++ cups-1.4rc1/templates/ja/modify-printer.tmpl	2009-06-18 18:58:12.243635944 +0100
@@ -23,7 +23,7 @@
 </TR>
 <TR>
 <TH CLASS="label">共有:</TH>
-<TD><INPUT TYPE="CHECKBOX" NAME="PRINTER_IS_SHARED" {PRINTER_IS_SHARED=1?CHECKED:}">
+<TD><INPUT TYPE="CHECKBOX" NAME="PRINTER_IS_SHARED" {PRINTER_IS_SHARED=1?CHECKED:}>
 このプリンターを共有する</TD>
 </TR>
 <TR>
diff -up cups-1.4rc1/templates/modify-printer.tmpl.str3229 cups-1.4rc1/templates/modify-printer.tmpl
--- cups-1.4rc1/templates/modify-printer.tmpl.str3229	2008-09-20 04:33:59.000000000 +0100
+++ cups-1.4rc1/templates/modify-printer.tmpl	2009-06-18 18:58:12.246635665 +0100
@@ -23,7 +23,7 @@
 </TR>
 <TR>
 <TH CLASS="label">Sharing:</TH>
-<TD><INPUT TYPE="CHECKBOX" NAME="PRINTER_IS_SHARED" {PRINTER_IS_SHARED=1?CHECKED:}">
+<TD><INPUT TYPE="CHECKBOX" NAME="PRINTER_IS_SHARED" {PRINTER_IS_SHARED=1?CHECKED:}>
 Share This Printer</TD>
 </TR>
 <TR>
diff -up cups-1.4rc1/templates/pl/add-printer.tmpl.str3229 cups-1.4rc1/templates/pl/add-printer.tmpl
--- cups-1.4rc1/templates/pl/add-printer.tmpl.str3229	2009-02-25 16:48:35.000000000 +0000
+++ cups-1.4rc1/templates/pl/add-printer.tmpl	2009-06-18 18:58:12.261635679 +0100
@@ -29,7 +29,7 @@
 </TR>
 <TR>
 <TH CLASS="label">Współdzielenie:</TH>
-<TD><INPUT TYPE="CHECKBOX" NAME="PRINTER_IS_SHARED" {PRINTER_IS_SHARED=1?CHECKED:}">
+<TD><INPUT TYPE="CHECKBOX" NAME="PRINTER_IS_SHARED" {PRINTER_IS_SHARED=1?CHECKED:}>
 Współdzielenie tej drukarki</TD>
 </TR>
 <TR>
diff -up cups-1.4rc1/templates/pl/modify-printer.tmpl.str3229 cups-1.4rc1/templates/pl/modify-printer.tmpl
--- cups-1.4rc1/templates/pl/modify-printer.tmpl.str3229	2009-02-25 16:48:35.000000000 +0000
+++ cups-1.4rc1/templates/pl/modify-printer.tmpl	2009-06-18 18:58:12.264511505 +0100
@@ -23,7 +23,7 @@
 </TR>
 <TR>
 <TH CLASS="label">Współdzielenie:</TH>
-<TD><INPUT TYPE="CHECKBOX" NAME="PRINTER_IS_SHARED" {PRINTER_IS_SHARED=1?CHECKED:}">
+<TD><INPUT TYPE="CHECKBOX" NAME="PRINTER_IS_SHARED" {PRINTER_IS_SHARED=1?CHECKED:}>
 Share This Printer</TD>
 </TR>
 <TR>
diff -up cups-1.4rc1/templates/ru/add-printer.tmpl.str3229 cups-1.4rc1/templates/ru/add-printer.tmpl
--- cups-1.4rc1/templates/ru/add-printer.tmpl.str3229	2009-02-17 19:04:25.000000000 +0000
+++ cups-1.4rc1/templates/ru/add-printer.tmpl	2009-06-18 18:58:12.249511736 +0100
@@ -29,7 +29,7 @@
 </TR>
 <TR>
 <TH CLASS="label">Совместный доступ:</TH>
-<TD><INPUT TYPE="CHECKBOX" NAME="PRINTER_IS_SHARED" {PRINTER_IS_SHARED=1?CHECKED:}">
+<TD><INPUT TYPE="CHECKBOX" NAME="PRINTER_IS_SHARED" {PRINTER_IS_SHARED=1?CHECKED:}>
 Разрешить совместный доступ к этому принтеру</TD>
 </TR>
 <TR>
diff -up cups-1.4rc1/templates/ru/modify-printer.tmpl.str3229 cups-1.4rc1/templates/ru/modify-printer.tmpl
--- cups-1.4rc1/templates/ru/modify-printer.tmpl.str3229	2009-02-17 19:04:25.000000000 +0000
+++ cups-1.4rc1/templates/ru/modify-printer.tmpl	2009-06-18 18:58:12.252511667 +0100
@@ -23,7 +23,7 @@
 </TR>
 <TR>
 <TH CLASS="label">Совместный доступ:</TH>
-<TD><INPUT TYPE="CHECKBOX" NAME="PRINTER_IS_SHARED" {PRINTER_IS_SHARED=1?CHECKED:}">
+<TD><INPUT TYPE="CHECKBOX" NAME="PRINTER_IS_SHARED" {PRINTER_IS_SHARED=1?CHECKED:}>
 Разрешить совместный доступ к этому принтеру</TD>
 </TR>
 <TR>
