diff -up cups-1.4rc1/filter/bannertops.c.str3277 cups-1.4rc1/filter/bannertops.c
--- cups-1.4rc1/filter/bannertops.c.str3277	2009-05-07 00:18:01.000000000 +0100
+++ cups-1.4rc1/filter/bannertops.c	2009-08-04 13:18:17.626076410 +0100
@@ -593,7 +593,7 @@ write_banner(banner_file_t *banner,	/* I
       printf("%.1f %.1f translate\n", PageLeft, PageBottom);
     else
       printf("%.1f %.1f translate\n", PageWidth - PageRight,
-             PageLength - PageRight);
+             PageLength - PageTop);
     puts("0 setgray");
 
     y = info_top;
