diff -up cups-1.4rc1/scheduler/printers.c.str3308 cups-1.4rc1/scheduler/printers.c
--- cups-1.4rc1/scheduler/printers.c.str3308	2009-08-26 18:16:22.948769384 +0100
+++ cups-1.4rc1/scheduler/printers.c	2009-08-26 18:16:23.649770491 +0100
@@ -3718,10 +3718,6 @@ add_printer_formats(cupsd_printer_t *p)	
                       p->name, mimetype);
   }
 
-  cupsdLogMessage(CUPSD_LOG_DEBUG2,
-                  "add_printer_formats: %s: %d supported types",
-		  p->name, cupsArrayCount(p->filetypes) + 1);
-
  /*
   * Add the file formats that can be filtered...
   */
@@ -3732,9 +3728,13 @@ add_printer_formats(cupsd_printer_t *p)	
   else
     i = 0;
 
+  cupsdLogMessage(CUPSD_LOG_DEBUG2,
+                  "add_printer_formats: %s: %d supported types",
+                  p->name, cupsArrayCount(p->filetypes) + i);
+
   attr = ippAddStrings(p->attrs, IPP_TAG_PRINTER, IPP_TAG_MIMETYPE,
                        "document-format-supported",
-		       cupsArrayCount(p->filetypes) + 1, NULL, NULL);
+                       cupsArrayCount(p->filetypes) + i, NULL, NULL);
 
   if (i)
     attr->values[0].string.text = _cupsStrAlloc("application/octet-stream");
