diff --git a/scheduler/job.c b/scheduler/job.c
index 49e4fa337..abe43e5ed 100644
--- a/scheduler/job.c
+++ b/scheduler/job.c
@@ -3165,6 +3165,149 @@ finalize_job(cupsd_job_t *job,		/* I - Job */
   snprintf(buffer, sizeof(buffer), "total %d", ippGetInteger(job->impressions, 0));
   cupsdLogPage(job, buffer);
 
+
+  /*
+   * Custom final impression (page) count...
+   */
+  int impressions = ippGetInteger(job->impressions, 0);
+  static char *cups_custom_impression_count_command = "/usr/local/libexec/cups_custom_impression_count";
+  if (access(cups_custom_impression_count_command, X_OK) == 0 && impressions > 0)
+  {
+    char *argv[10];           /* Command-line arguments */
+    char *envp[MAX_ENV];      /* Environment */
+    char * username;
+    int   color_supported;
+    char  number[256];        /* Page number */
+    int   copies;             /* Number of copies */
+    int   temppid;            /* Process ID for custom impression count process */
+    int   outfds[2];          /* Pipe file descriptors for stdout */
+    int   errfds[2];          /* Pipe file descriptors for stderr output */
+    ipp_attribute_t *attr;
+
+    strlcpy(number, "1", sizeof(number));
+    copies = 1;
+    sscanf(buffer, "%255s%d", number, &copies);
+
+    color_supported = 0;
+    attr = ippFindAttribute(job->printer->attrs, "color-supported", IPP_TAG_BOOLEAN);
+    if (attr)
+    {
+      color_supported = ippGetBoolean(attr, 0);
+    }
+
+    strlcpy(number, "1", sizeof(number));
+    copies = 1;
+    sscanf(buffer, "%255s%d", number, &copies);
+
+    /* Page number */
+    number;
+
+    /* Number of copies */
+    copies;
+
+    /* Printer name */
+    job->dest;
+
+
+    attr = ippFindAttribute(job->attrs, "media", IPP_TAG_BEGIN_COLLECTION);
+    if (attr)
+    {
+      color_supported = ippGetBoolean(attr, 0);
+    }
+
+
+   argv[0] = cups_custom_impression_count_command;
+   argv[1] = color_supported ? "1" : "0"; /* color supported */
+   argv[2] = job->username ? job->username : "-"; /* Username */
+   argv[3] = number;
+   argv[4] = 
+   argv[5] = owner ? (char *)owner : "";
+   argv[6] = req_username;
+   argv[7] = orig_username;
+   argv[8] = accounting_id;
+   argv[9] = NULL;
+
+   cupsdLoadEnv(envp, (int)(sizeof(envp) / sizeof(envp[0])));
+
+   if (cupsdOpenPipe(outfds) || cupsdOpenPipe(errfds))
+   {
+     cupsdLogMessage(CUPSD_LOG_ERROR, "Unable to create pipe for %s - %s",
+                     cups_custom_auth_command, strerror(errno));
+     return (HTTP_UNAUTHORIZED);
+   }
+
+   if(cupsdStartProcess(cups_custom_auth_command, argv, envp, -1, outfds[1], errfds[1], -1, -1, 0, DefaultProfile, NULL, &temppid) < 0)
+   {
+     /*
+      * Error - can't fork!
+      */
+
+     cupsdLogMessage(CUPSD_LOG_ERROR, "Unable to fork %s - %s",
+                     cups_custom_auth_command, strerror(errno));
+
+     cupsdClosePipe(outfds);
+     cupsdClosePipe(errfds);
+     return (HTTP_UNAUTHORIZED);
+   }
+   else
+   {
+     /*
+      * Fork successful ...
+      */
+     char  buf[2048];
+     FILE* outfp = fdopen(outfds[0], "r");
+     FILE* errfp = fdopen(errfds[0], "r");
+
+     cupsdLogMessage(CUPSD_LOG_DEBUG, "Started %s- PID = %d", cups_custom_auth_command, temppid);
+
+     close(outfds[1]);
+     close(errfds[1]);
+
+     buf[0] = '\0';
+     while (fgets(buf, sizeof(buf), errfp) != NULL)
+     {
+       buf[strcspn(buf, "\r\n")] = '\0';      /* chomp any kind of newline */
+       if (strncmp(buf, "DEBUG:", 6) == 0) {
+         cupsdLogMessage(CUPSD_LOG_DEBUG, "cupsdIsAuthorized custom authorization:%s", &buf[6]);
+       } else if (strncmp(buf, "WARN:", 5) == 0) {
+         cupsdLogMessage(CUPSD_LOG_WARN, "cupsdIsAuthorized custom authorization:%s", &buf[5]);
+       } else if (strncmp(buf, "ERROR:", 6) == 0) {
+         cupsdLogMessage(CUPSD_LOG_ERROR, "cupsdIsAuthorized custom authorization:%s", &buf[6]);
+       } else {
+         cupsdLogMessage(CUPSD_LOG_ERROR, "cupsdIsAuthorized custom authorization: %s", buf);
+       }
+     }
+
+     username[0] = '\0';
+     while (fgets(buf, sizeof(buf), outfp) != NULL)
+     {
+       buf[strcspn(buf, "\r\n")] = '\0';      /* chomp any kind of newline */
+       if (strncmp(buf, "username=", 9) == 0) {
+         cupsdLogMessage(CUPSD_LOG_DEBUG, "cupsdIsAuthorized using custom authorization: username=\"%s\"",
+                         &buf[9]);
+         strlcpy(username, &buf[9], sizeof(username));
+         strlcpy(con->username, username, sizeof(con->username));
+         /* replace requesting-user-name with custom authorization username */
+         if (strncmp(req_username, username, sizeof(username) != 0))
+         {
+           attr = ippFindAttribute(con->request, "requesting-user-name", IPP_TAG_NAME);
+           if (attr)
+           {
+             if (attr->values[0].string.text) _cupsStrFree(attr->values[0].string.text);
+             attr->values[0].string.text = _cupsStrAlloc(username);
+           }
+         }
+
+       } else if (strncmp(buf, "auth-type=", 10) == 0) {
+         cupsdLogMessage(CUPSD_LOG_WARN, "cupsdIsAuthorized custom authorization auth-type:%s", &buf[10]);
+         /* TO-DO, overide authentication type with custom authentication type */
+       }
+     }
+
+     close(outfds[0]);
+     close(errfds[0]);
+  }
+
  /*
   * Process the exit status...
   */
