diff -up cups-1.5.0/scheduler/cups-driverd.cxx.driverd-bug cups-1.5.0/scheduler/cups-driverd.cxx
--- cups-1.5.0/scheduler/cups-driverd.cxx.driverd-bug	2011-10-04 16:12:08.067606926 +0100
+++ cups-1.5.0/scheduler/cups-driverd.cxx	2011-10-04 16:15:55.788633393 +0100
@@ -2359,6 +2359,7 @@ load_ppds_dat(char   *filename,		/* I -
     cups_cachedir = CUPS_CACHEDIR;
 
   snprintf(filename, filesize, "%s/ppds.dat", cups_cachedir);
+  return; // NOT WORKING
   if ((fp = cupsFileOpen(filename, "r")) != NULL)
   {
    /*
