diff -up cups-1.3.2/scheduler/dirsvc.c.str2512 cups-1.3.2/scheduler/dirsvc.c
--- cups-1.3.2/scheduler/dirsvc.c.str2512	2007-09-21 13:46:23.000000000 +0100
+++ cups-1.3.2/scheduler/dirsvc.c	2007-09-21 13:46:54.000000000 +0100
@@ -912,6 +912,7 @@ cupsdSendBrowseList(void)
         cupsArraySave(Printers);
         cupsdDeletePrinter(p, 1);
         cupsArrayRestore(Printers);
+        cupsdWritePrintcap();
       }
     }
   }
