diff -up cups-1.7rc1/cups/http.c.build cups-1.7rc1/cups/http.c
--- cups-1.7rc1/cups/http.c.build	2013-07-22 14:51:18.673065709 +0100
+++ cups-1.7rc1/cups/http.c	2013-07-22 14:51:19.781070570 +0100
@@ -5247,7 +5247,7 @@ http_setup_ssl(http_t *http)		/* I - Con
     if (!message)
       message = _("Unable to establish a secure connection to host.");
 
-    _cupsSetError(IPP_PKI_ERROR, message, 1);
+    _cupsSetError(IPP_STATUS_ERROR_CUPS_PKI, message, 1);
 
     return (-1);
   }
