diff -up cups-1.7.1/cups/request.c.str4386 cups-1.7.1/cups/request.c
--- cups-1.7.1/cups/request.c.str4386	2013-07-23 13:33:52.000000000 +0100
+++ cups-1.7.1/cups/request.c	2014-03-06 10:58:39.018807898 +0000
@@ -261,7 +261,7 @@ cupsDoIORequest(http_t     *http,	/* I -
     * Get the server's response...
     */
 
-    if (status != HTTP_STATUS_ERROR)
+    if (status <= HTTP_STATUS_CONTINUE)
     {
       response = cupsGetResponse(http, resource);
       status   = httpGetStatus(http);
