diff -up cups-2.1rc1/scheduler/log.c.journal cups-2.1rc1/scheduler/log.c
--- cups-2.1rc1/scheduler/log.c.journal	2015-08-10 15:06:15.000000000 +0200
+++ cups-2.1rc1/scheduler/log.c	2015-08-10 15:07:30.046368863 +0200
@@ -690,7 +690,7 @@ cupsdLogJob(cupsd_job_t *job,		/* I - Jo
 
     va_end(ap);
 
-    if (job)
+    if (job && printer)
       sd_journal_send("MESSAGE=%s", log_line,
 		      "PRIORITY=%i", log_levels[level],
 		      PWG_Event"=JobStateChanged",
