diff -up cups-1.7.4/scheduler/client.c~ cups-1.7.4/scheduler/client.c
--- cups-1.7.4/scheduler/client.c~	2014-07-18 13:34:25.243248601 +0100
+++ cups-1.7.4/scheduler/client.c	2014-07-18 14:50:55.356614243 +0100
@@ -3980,12 +3980,7 @@ pipe_command(cupsd_client_t *con,	/* I -
   argv[0] = command;
 
   if (options)
-  {
-    commptr = options;
-    if (*commptr == ' ')
-      commptr ++;
-    strlcpy(argbuf, commptr, sizeof(argbuf));
-  }
+    strlcpy(argbuf, options, sizeof(argbuf));
   else
     argbuf[0] = '\0';
 
