--- cups-1.1.21rc2/systemv/Makefile.noinit	2004-08-11 15:10:22.000000000 +0100
+++ cups-1.1.21rc2/systemv/Makefile	2004-08-24 12:30:31.868411346 +0100
@@ -77,7 +77,7 @@
 	$(INSTALL_BIN) lp $(BINDIR)
 	$(INSTALL_BIN) lpoptions $(BINDIR)
 	$(INSTALL_BIN) lpstat $(BINDIR)
-	-$(INSTALL_BIN) -m 4755 -o $(CUPS_USER) -g $(CUPS_GROUP) lppasswd $(BINDIR)
+	-$(INSTALL_BIN) -m 4755 lppasswd $(BINDIR)
 	if test ! -x $(BINDIR)/lppasswd; then \
 		chmod 755 $(BINDIR)/lppasswd; \
 	fi
--- cups-1.1.21rc2/Makefile.noinit	2004-02-25 20:14:50.000000000 +0000
+++ cups-1.1.21rc2/Makefile	2004-08-24 12:28:17.704029413 +0100
@@ -89,30 +89,6 @@
 	echo Installing cups-config script...
 	$(INSTALL_DIR) $(BINDIR)
 	$(INSTALL_SCRIPT) cups-config $(BINDIR)/cups-config
-	echo Installing startup script...
-	if test "x$(INITDIR)" != "x"; then \
-		$(INSTALL_DIR) $(BUILDROOT)$(INITDIR)/init.d; \
-		$(INSTALL_SCRIPT) cups.sh $(BUILDROOT)$(INITDIR)/init.d/cups; \
-		$(INSTALL_DIR) $(BUILDROOT)$(INITDIR)/rc0.d; \
-		$(INSTALL_SCRIPT) cups.sh  $(BUILDROOT)$(INITDIR)/rc0.d/K00cups; \
-		$(INSTALL_DIR) $(BUILDROOT)$(INITDIR)/rc2.d; \
-		$(INSTALL_SCRIPT) cups.sh $(BUILDROOT)$(INITDIR)/rc2.d/S99cups; \
-		$(INSTALL_DIR) $(BUILDROOT)$(INITDIR)/rc3.d; \
-		$(INSTALL_SCRIPT) cups.sh $(BUILDROOT)$(INITDIR)/rc3.d/S99cups; \
-		$(INSTALL_DIR) $(BUILDROOT)$(INITDIR)/rc5.d; \
-		$(INSTALL_SCRIPT) cups.sh $(BUILDROOT)$(INITDIR)/rc5.d/S99cups; \
-	fi
-	if test "x$(INITDIR)" = "x" -a "x$(INITDDIR)" != "x"; then \
-		$(INSTALL_DIR) $(BUILDROOT)$(INITDDIR); \
-		if test "$(INITDDIR)" = "/System/Library/StartupItems/PrintingServices"; then \
-			$(INSTALL_SCRIPT) cups.osx $(BUILDROOT)$(INITDDIR)/PrintingServices; \
-			$(INSTALL_DATA) cups.plist $(BUILDROOT)$(INITDDIR)/StartupParameters.plist; \
-			$(INSTALL_DIR) $(BUILDROOT)$(INITDDIR)/Resources/English.lproj; \
-			$(INSTALL_DATA) cups.strings $(BUILDROOT)$(INITDDIR)/Resources/English.lproj/Localizable.strings; \
-		else \
-			$(INSTALL_SCRIPT) cups.sh $(BUILDROOT)$(INITDDIR)/cups; \
-		fi \
-	fi
 
 #
 # Install source and header files...
--- cups-1.1.21rc2/Makedefs.in.noinit	2004-06-29 04:14:19.000000000 +0100
+++ cups-1.1.21rc2/Makedefs.in	2004-08-24 12:28:17.713027695 +0100
@@ -47,7 +47,7 @@
 # Installation programs...
 #
 
-INSTALL_BIN	=	$(LIBTOOL) $(INSTALL) -m 755 -s
+INSTALL_BIN	=	$(LIBTOOL) $(INSTALL) -m 755
 INSTALL_DATA	=	$(INSTALL) -m 644
 INSTALL_DIR	=	$(INSTALL) -d
 INSTALL_LIB	=	$(LIBTOOL) $(INSTALL) -m 755
