diff -ruN cups-1.1.19.orig/cups/language.c cups-1.1.19/cups/language.c
--- cups-1.1.19.orig/cups/language.c	2003-08-14 20:13:18.000000000 +0900
+++ cups-1.1.19/cups/language.c	2003-08-14 20:37:37.000000000 +0900
@@ -188,6 +188,7 @@
 			charset[16],	/* Character set */
 			*ptr,		/* Pointer into language/ */
 			real[48],	/* Real language name */
+			ll_cc[48],	/* Language + Country name */
 			filename[1024],	/* Filename for language locale file */
 			*localedir;	/* Directory for locale files */
   cups_encoding_t	encoding;	/* Encoding to use */
@@ -356,6 +357,7 @@
   */
 
   snprintf(real, sizeof(real), "%s_%s", langname, country);
+  strcpy(ll_cc, real);
 
   if ((lang = cups_cache_lookup(real, encoding)) != NULL)
     return (lang);
@@ -468,7 +470,7 @@
   */
 
   lang->used ++;
-  strlcpy(lang->language, real, sizeof(lang->language));
+  strlcpy(lang->language, ll_cc, sizeof(lang->language));
 
   if (charset[0])
     lang->encoding = encoding;
