diff -up cups-1.4b2/cups/encode.c.str3077 cups-1.4b2/cups/encode.c
--- cups-1.4b2/cups/encode.c.str3077	2008-12-15 17:54:11.000000000 +0000
+++ cups-1.4b2/cups/encode.c	2009-01-28 16:58:01.000000000 +0000
@@ -69,10 +69,10 @@ static const _ipp_option_t ipp_options[]
   { 0, "hue-default",		IPP_TAG_INTEGER,	IPP_TAG_PRINTER },
   { 1, "include-schemes",	IPP_TAG_NAME,		IPP_TAG_OPERATION },
   { 0, "job-impressions",	IPP_TAG_INTEGER,	IPP_TAG_JOB },
-  { 0, "job-k-limit",		IPP_TAG_INTEGER,	IPP_TAG_JOB },
-  { 0, "job-page-limit",	IPP_TAG_INTEGER,	IPP_TAG_JOB },
+  { 0, "job-k-limit",		IPP_TAG_INTEGER,	IPP_TAG_PRINTER },
+  { 0, "job-page-limit",	IPP_TAG_INTEGER,	IPP_TAG_PRINTER },
   { 0, "job-priority",		IPP_TAG_INTEGER,	IPP_TAG_JOB },
-  { 0, "job-quota-period",	IPP_TAG_INTEGER,	IPP_TAG_JOB },
+  { 0, "job-quota-period",	IPP_TAG_INTEGER,	IPP_TAG_PRINTER },
   { 1, "job-sheets",		IPP_TAG_NAME,		IPP_TAG_JOB },
   { 1, "job-sheets-default",	IPP_TAG_NAME,		IPP_TAG_PRINTER },
   { 0, "job-uuid",		IPP_TAG_URI,		IPP_TAG_JOB },
