diff -up cups-2.1.4/backend/org.cups.usb-quirks.libusb-quirks cups-2.1.4/backend/org.cups.usb-quirks
--- cups-2.1.4/backend/org.cups.usb-quirks.libusb-quirks	2016-06-15 14:46:46.122497874 +0200
+++ cups-2.1.4/backend/org.cups.usb-quirks	2016-06-15 14:48:47.752552749 +0200
@@ -76,6 +76,9 @@
 # Canon, Inc. MP510 Printer (https://bugs.launchpad.net/bugs/1050009)
 0x04a9 0x1717 unidir
 
+# Canon, Inc. MP540 Printer, https://bugzilla.redhat.com/967873
+0x04a9 0x1730 unidir
+
 # Canon, Inc. MP550 Printer (Issue #4155)
 0x04a9 0x173d unidir
 
