diff -up cups-1.4b2/CHANGES.txt.local-protocols cups-1.4b2/CHANGES.txt
diff -up cups-1.4b2/conf/cupsd.conf.in.local-protocols cups-1.4b2/conf/cupsd.conf.in
--- cups-1.4b2/conf/cupsd.conf.in.local-protocols	2009-01-09 16:47:15.000000000 +0000
+++ cups-1.4b2/conf/cupsd.conf.in	2009-01-27 14:46:59.000000000 +0000
@@ -23,7 +23,7 @@ Listen localhost:@DEFAULT_IPP_PORT@
 Browsing On
 BrowseOrder allow,deny
 BrowseAllow all
-BrowseLocalProtocols @CUPS_DEFAULT_BROWSE_LOCAL_PROTOCOLS@
+BrowseLocalProtocols @CUPS_BROWSE_LOCAL_PROTOCOLS@
 
 # Default authentication type, when authentication is required...
 DefaultAuthType Basic
