diff -up cups-2.2.6/scheduler/log.c.moved-logs cups-2.2.6/scheduler/log.c
--- cups-2.2.6/scheduler/log.c.moved-logs	2018-02-06 14:24:32.857387592 +0100
+++ cups-2.2.6/scheduler/log.c	2018-02-06 14:24:32.917387049 +0100
@@ -979,7 +979,7 @@ cupsdLogPage(cupsd_job_t *job,		/* I - J
   *bufptr = '\0';
 
 #ifdef HAVE_SYSTEMD_SD_JOURNAL_H
-  if (!strcmp(ErrorLog, "syslog"))
+  if (!strcmp(PageLog, "syslog"))
   {
     static const char * const job_states[] =
     {					/* job-state strings */
@@ -1176,7 +1176,7 @@ cupsdLogRequest(cupsd_client_t *con,	/*
   }
 
 #ifdef HAVE_SYSTEMD_SD_JOURNAL_H
-  if (!strcmp(ErrorLog, "syslog"))
+  if (!strcmp(AccessLog, "syslog"))
   {
     sd_journal_print(LOG_INFO, "REQUEST %s - %s \"%s %s HTTP/%d.%d\" %d " CUPS_LLFMT " %s %s", con->http->hostname, con->username[0] != '\0' ? con->username : "-", states[con->operation], _httpEncodeURI(temp, con->uri, sizeof(temp)), con->http->version / 100, con->http->version % 100, code, CUPS_LLCAST con->bytes, con->request ? ippOpString(con->request->request.op.operation_id) : "-", con->response ? ippErrorString(con->response->request.status.status_code) : "-");
     return (1);
