--- cups-1.3b1/config-scripts/cups-dnssd.m4.avahi	2007-07-19 13:45:14.000000000 +0100
+++ cups-1.3b1/config-scripts/cups-dnssd.m4	2007-07-19 13:45:55.000000000 +0100
@@ -37,6 +37,9 @@
 				AC_DEFINE(HAVE_COREFOUNDATION)
 				AC_DEFINE(HAVE_SYSTEMCONFIGURATION)
 				;;
+			Linux*)
+				DNSSDLIBS="-ldns_sd"
+				;;
 			*)
 				# All others...
 				DNSSDLIBS="???"
--- cups-1.3b1/scheduler/dirsvc.c.avahi	2007-07-19 13:43:51.000000000 +0100
+++ cups-1.3b1/scheduler/dirsvc.c	2007-07-19 13:44:42.000000000 +0100
@@ -65,6 +65,8 @@
 
 #ifdef HAVE_DNSSD
 #  include <dns_sd.h>
+#endif /* HAVE_DNSSD */
+#ifdef __APPLE__
 #  include <nameser.h>
 #  include <nameser.h>
 #  ifdef HAVE_COREFOUNDATION
@@ -73,7 +75,7 @@
 #  ifdef HAVE_SYSTEMCONFIGURATION
 #    include <SystemConfiguration/SystemConfiguration.h>
 #  endif /* HAVE_SYSTEMCONFIGURATION */
-#endif /* HAVE_DNSSD */
+#endif /* __APPLE__ */
 
 
 /*
--- cups-1.3b1/configure.avahi	2007-07-19 13:46:30.000000000 +0100
+++ cups-1.3b1/configure	2007-07-19 13:46:57.000000000 +0100
@@ -17294,6 +17294,9 @@
 _ACEOF
 
 				;;
+			Linux*)
+				DNSSDLIBS="-ldns_sd"
+				;;
 			*)
 				# All others...
 				DNSSDLIBS="???"
