diff --git a/cups/ipp-support.c b/cups/ipp-support.c
index bb0c6c8bd..b0257b085 100644
--- a/cups/ipp-support.c
+++ b/cups/ipp-support.c
@@ -1671,6 +1671,8 @@ ippCreateRequestedArray(ipp_t *request)	/* I - IPP request */
     "job-page-limit",			/* CUPS extension */
     "job-password-encryption-supported",
     "job-password-supported",
+    "job-password-repertoire-configured",
+    "job-password-repertoire-supported",
     "job-presets-supported",		/* IPP Presets */
     "job-privacy-attributes",		/* IPP Privacy Attributes */
     "job-privacy-scope",		/* IPP Privacy Attributes */
diff --git a/scheduler/printers.c b/scheduler/printers.c
index 5f9852e64..115aacd8c 100644
--- a/scheduler/printers.c
+++ b/scheduler/printers.c
@@ -4081,6 +4081,52 @@ load_ppd(cupsd_printer_t *p)		/* I - Printer */
                    "job-password-encryption-supported", NULL, "none");
       ippAddInteger(p->ppd_attrs, IPP_TAG_PRINTER, IPP_TAG_INTEGER,
                     "job-password-supported", (int)strlen(p->pc->password));
+
+      switch (p->pc->password[0])
+      {
+        case '1' :
+          ippAddString(p->ppd_attrs, IPP_TAG_PRINTER, IPP_TAG_KEYWORD,
+                      "job-password-repertoire-configured", NULL, "iana_us-ascii_digits");
+          ippAddString(p->ppd_attrs, IPP_TAG_PRINTER, IPP_TAG_KEYWORD,
+                      "job-password-repertoire-supported", NULL, "iana_us-ascii_digits");
+          break;
+        case 'A' :
+          ippAddString(p->ppd_attrs, IPP_TAG_PRINTER, IPP_TAG_KEYWORD,
+                      "job-password-repertoire-configured", NULL, "iana_us-ascii_letters");
+          ippAddString(p->ppd_attrs, IPP_TAG_PRINTER, IPP_TAG_KEYWORD,
+                      "job-password-repertoire-supported", NULL, "iana_us-ascii_letters");
+          break;
+        case 'C' :
+          ippAddString(p->ppd_attrs, IPP_TAG_PRINTER, IPP_TAG_KEYWORD,
+                      "job-password-repertoire-configured", NULL, "iana_us-ascii_complex");
+          ippAddString(p->ppd_attrs, IPP_TAG_PRINTER, IPP_TAG_KEYWORD,
+                      "job-password-repertoire-supported", NULL, "iana_us-ascii_complex");
+          break;
+        case '.' :
+          ippAddString(p->ppd_attrs, IPP_TAG_PRINTER, IPP_TAG_KEYWORD,
+                      "job-password-repertoire-configured", NULL, "iana_us-ascii_any");
+          ippAddString(p->ppd_attrs, IPP_TAG_PRINTER, IPP_TAG_KEYWORD,
+                      "job-password-repertoire-supported", NULL, "iana_us-ascii_any");
+          break;
+        case 'N' :
+          ippAddString(p->ppd_attrs, IPP_TAG_PRINTER, IPP_TAG_KEYWORD,
+                      "job-password-repertoire-configured", NULL, "iana_utf-8_digits");
+          ippAddString(p->ppd_attrs, IPP_TAG_PRINTER, IPP_TAG_KEYWORD,
+                      "job-password-repertoire-supported", NULL, "iana_utf-8_digits");
+          break;
+        case 'U' :
+          ippAddString(p->ppd_attrs, IPP_TAG_PRINTER, IPP_TAG_KEYWORD,
+                      "job-password-repertoire-configured", NULL, "iana_utf-8_letters");
+          ippAddString(p->ppd_attrs, IPP_TAG_PRINTER, IPP_TAG_KEYWORD,
+                      "job-password-repertoire-supported", NULL, "iana_utf-8_letters");
+          break;
+        case '*' :
+          ippAddString(p->ppd_attrs, IPP_TAG_PRINTER, IPP_TAG_KEYWORD,
+                      "job-password-repertoire-configured", NULL, "iana_utf-8_any");
+          ippAddString(p->ppd_attrs, IPP_TAG_PRINTER, IPP_TAG_KEYWORD,
+                      "job-password-repertoire-supported", NULL, "iana_utf-8_any");
+          break;
+      }
     }
 
     if (ppd->throughput)
