diff -up cups-1.4rc1/scheduler/ipp.c.str3258 cups-1.4rc1/scheduler/ipp.c
--- cups-1.4rc1/scheduler/ipp.c.str3258	2009-07-15 09:58:15.559299247 +0100
+++ cups-1.4rc1/scheduler/ipp.c	2009-07-15 09:58:47.981299548 +0100
@@ -1721,8 +1721,7 @@ add_job(cupsd_client_t  *con,		/* I - Cl
     attr = ippAddString(job->attrs, IPP_TAG_JOB, IPP_TAG_KEYWORD,
                         "job-hold-until", NULL, val);
   }
-  if (attr && strcmp(attr->values[0].string.text, "no-hold") &&
-      !(printer->type & CUPS_PRINTER_REMOTE))
+  if (attr && strcmp(attr->values[0].string.text, "no-hold"))
   {
    /*
     * Hold job until specified time...
diff -up cups-1.4rc1/scheduler/job.c.str3258 cups-1.4rc1/scheduler/job.c
--- cups-1.4rc1/scheduler/job.c.str3258	2009-07-15 09:58:15.627173602 +0100
+++ cups-1.4rc1/scheduler/job.c	2009-07-15 09:58:47.984298957 +0100
@@ -3095,7 +3095,8 @@ get_options(cupsd_job_t *job,		/* I - Jo
 	  attr->value_tag == IPP_TAG_BEGIN_COLLECTION) /* Not yet supported */
 	continue;
 
-      if (!strncmp(attr->name, "time-", 5))
+      if (!strncmp(attr->name, "time-", 5) ||
+          !strcmp(attr->name, "job-hold-until"))
 	continue;
 
       if (!strncmp(attr->name, "job-", 4) &&
