--- cups-1.2.11/backend/usb-unix.c.direct-usb	2007-05-04 15:03:02.000000000 +0100
+++ cups-1.2.11/backend/usb-unix.c	2007-06-13 16:46:56.000000000 +0100
@@ -101,6 +101,9 @@
              strcasecmp(hostname, "Minolta");
 #endif /* __FreeBSD__ || __NetBSD__ || __OpenBSD__ || __DragonFly__ */
 
+    if (use_bc && !strncmp(uri, "usb:/dev/", 9))
+      use_bc = 0;
+
     if ((device_fd = open_device(uri, &use_bc)) == -1)
     {
       if (getenv("CLASS") != NULL)
@@ -317,12 +320,7 @@
   if (!strncmp(uri, "usb:/dev/", 9))
 #ifdef __linux
   {
-   /*
-    * Do not allow direct devices anymore...
-    */
-
-    errno = ENODEV;
-    return (-1);
+    return (open(uri + 4, O_RDWR | O_EXCL));
   }
   else if (!strncmp(uri, "usb://", 6))
   {
