diff --git a/backend/ipp.c b/backend/ipp.c
index 8609e3a28..10b1ff0b2 100644
--- a/backend/ipp.c
+++ b/backend/ipp.c
@@ -115,6 +115,7 @@ static const char * const pattrs[] =	/* Printer attributes we want */
   "printer-alert",
   "printer-alert-description",
   "printer-is-accepting-jobs",
+  "printer-make-and-model",
   "printer-mandatory-job-attributes",
   "printer-state",
   "printer-state-message",
@@ -254,6 +255,7 @@ main(int  argc,				/* I - Number of command-line args */
   ipp_attribute_t *printer_accepting;	/* printer-is-accepting-jobs */
   ipp_attribute_t *print_color_mode_sup;/* Does printer support print-color-mode? */
   ipp_attribute_t *print_scaling_sup;	/* print-scaling-supported */
+  ipp_attribute_t *make_and_model;      /* printer-make-and-model */
   int		create_job = 0,		/* Does printer support Create-Job? */
 		get_job_attrs = 0,	/* Does printer support Get-Job-Attributes? */
 		send_document = 0,	/* Does printer support Send-Document? */
@@ -1955,8 +1957,15 @@ main(int  argc,				/* I - Number of command-line args */
       */
 
       fputs("JOBSTATE: cups-retry-as-raster\n", stderr);
-      if (job_id > 0)
-	cancel_job(http, uri, job_id, resource, argv[2], version);
+      if (job_id > 0) {
+        if ((make_and_model = ippFindAttribute(response, "printer-make-and-model", IPP_TAG_TEXT)) != NULL)
+        {
+          if (strcmp(make_and_model->values[0].string.text, "KONICA MINOLTA bizhub C458") != 0) {
+            fprintf(stderr, "DEBUG: KONICA MINOLTA bizhub C458: cups-retry-as-raster\n");
+          }
+        }
+        cancel_job(http, uri, job_id, resource, argv[2], version);
+      }
 
       goto cleanup;
     }
@@ -2525,6 +2534,7 @@ monitor_printer(
   ipp_t		*request,		/* IPP request */
 		*response;		/* IPP response */
   ipp_attribute_t *attr;		/* Attribute in response */
+  ipp_attribute_t *make_and_model;      /* printer-make-and-model */
   int		delay,			/* Current delay */
 		prev_delay;		/* Previous delay */
   ipp_op_t	job_op;			/* Operation to use */
@@ -2793,6 +2803,12 @@ monitor_printer(
 
       if (cupsLastError() > IPP_STATUS_OK_CONFLICTING)
       {
+        if ((make_and_model = ippFindAttribute(response, "printer-make-and-model", IPP_TAG_TEXT)) != NULL)
+        {
+          if (strcmp(make_and_model->values[0].string.text, "KONICA MINOLTA bizhub C458") != 0) {
+            fprintf(stderr, "DEBUG: KONICA MINOLTA bizhub C458:cancel_job() = %s\n", cupsLastErrorString());
+          }
+        }
 	fprintf(stderr, "DEBUG: (monitor) cancel_job() = %s\n", cupsLastErrorString());
 	_cupsLangPrintFilter(stderr, "ERROR", _("Unable to cancel print job."));
       }
