--- cups-1.2.4/conf/cupsd.conf.in.directed-broadcast	2006-10-05 12:33:08.000000000 +0100
+++ cups-1.2.4/conf/cupsd.conf.in	2006-10-05 12:33:04.000000000 +0100
@@ -21,6 +21,7 @@
 # Show shared printers on the local network.
 Browsing On
 BrowseOrder allow,deny
+# (Change '@LOCAL' to 'ALL' if using directed broadcasts from another subnet.)
 BrowseAllow @LOCAL
 
 # Default authentication type, when authentication is required...
--- cups-1.2.4/cups/adminutil.c.directed-broadcast	2006-10-05 12:16:48.000000000 +0100
+++ cups-1.2.4/cups/adminutil.c	2006-10-05 12:19:29.000000000 +0100
@@ -1155,7 +1155,11 @@
 	  cupsFilePuts(temp, "BrowseOrder allow,deny\n");
 
 	  if (remote_printers)
+	  {
+	    cupsFilePuts(temp, "# (Change '@LOCAL' to 'ALL' if using "
+	                       "directed broadcasts from another subnet.)\n");
 	    cupsFilePuts(temp, "BrowseAllow @LOCAL\n");
+	  }
 
 	  if (share_printers)
 	    cupsFilePuts(temp, "BrowseAddress @LOCAL\n");
